docs/dev/dockerfile-llb.md

```diff
@@ -0,0 +1,208 @@
+
+- Base image
+
+  ```dockerfile
+  FROM golang:1.12
+  ```
+
+  ```golang
+  st := llb.Image("golang:1.12")
+  ```
+
+- Scratch image
+
+  ```dockerfile
+  FROM scratch
+  ```
+
+  ```golang
+  st := llb.Scratch()
+  ```
+
+- Environment variables
+
+  ```dockerfile
+  ENV DEBIAN_FRONTEND=noninteractive
+  ```
+
+  ```golang
+  st = st.AddEnv("DEBIAN_FRONTEND", "noninteractive")
+  ```
+
+- Running programs
+
+  ```dockerfile
+  RUN echo hello
+  ```
+
+  ```golang
+  st = st.Run(
+    llb.Shlex("echo hello"),
+  ).Root()
+  ```
+
+- Working directory
+
+  ```dockerfile
+  WORKDIR /path
+  ```
+
+  ```golang
+  st = st.Dir("/path")
+  ```
+
+## File operations
+
+This is where LLB starts to deviate from Dockerfile in features. In
+Dockerfiles, the run command is completely opaque to the builder and just
+executes the command. But in LLB, there are file operations that have better
+caching semantics and understanding of the command:
+
+- Copying files
+
+  ```dockerfile
+  COPY --from=builder /files/* /files
+  ```
+
+  ```golang
+  var CopyOptions = &llb.CopyInfo{
+    FollowSymlinks:      true,
+    CopyDirContentsOnly: true,
+    AttemptUnpack:       false,
+    CreateDestPath:      true,
+    AllowWildcard:       true,
+    AllowEmptyWildcard:  true,
+  }
+  st = st.File(
+    llb.Copy(builder, "/files/*", "/files", CopyOptions),
+  )
+  ```
+
+- Adding files
+
+  ```dockerfile
+  ADD --from=builder /files.tgz /files
+  ```
+
+  ```golang
+  var AddOptions = &llb.CopyInfo{
+    FollowSymlinks:      true,
+    CopyDirContentsOnly: true,
+    AttemptUnpack:       true,
+    CreateDestPath:      true,
+    AllowWildcard:       true,
+    AllowEmptyWildcard:  true,
+  }
+  st = st.File(
+    llb.Copy(builder, "/files.tgz", "files", AddOptions),
+  )
+  ```
+
+- Chaining file commands
+
+  ```dockerfile
+  # not possible without RUN in Dockerfile
+  RUN mkdir -p /some && echo hello > /some/file
+  ```
+
+  ```golang
+  st = st.File(
+    llb.Mkdir("/some", 0755),
+  ).File(
+    llb.Mkfile("/some/file", 0644, "hello"),
+  )
+  ```
+
+## Bind mounts
+
+Bind mounts allow unidirectional syncing of the host's local file system into
+the build environment.
+
+Bind mounts in Buildkit should not be confused with bind mounts in the linux
+kernel - they do not sync bidirectionally. Bind mounts are only a snapshot of
+your local state, which is specified through the `llb.Local` state object:
+
+- Using bind mounts
+
+  ```dockerfile
+  WORKDIR /builder
+  RUN --mount=type=bind,target=/builder \
+  PIP_INDEX_URL=https://my-proxy.com/pypi \
+      pip install .
+  ```
+
+  ```golang
+  localState := llb.Local(
+    "context",
+    llb.SessionID(client.BuildOpts().SessionID),
+    llb.WithCustomName("loading .")
+    llb.FollowPaths([]string{"."}),
+  )
+
+  st = st.Dir("/builder").Run(
+    llb.Shlex("pip install ."),
+    llb.AddEnv(
+      "PIP_INDEX_URL",
+      "https://my-proxy.com/pypi",
+    ),
+    llb.AddMount("/builder", localState)
```

Comment:
This could show how to deal with outputs from RW mounts as that case is not possible in Dockerfiles.